io-engine/src/grpc/v0/nexus_grpc.rs

```diff
@@ -9,70 +10,60 @@ use crate::{
-            ChildState::Closed => rpc::ChildState::ChildDegraded,
-            ChildState::Faulted(reason) => match reason {
-                Reason::NoSpace => rpc::ChildState::ChildDegraded,
-                Reason::TimedOut => rpc::ChildState::ChildDegraded,
-                Reason::Unknown => rpc::ChildState::ChildFaulted,
-                Reason::CantOpen => rpc::ChildState::ChildFaulted,
-                Reason::RebuildFailed => rpc::ChildState::ChildFaulted,
-                Reason::IoError => rpc::ChildState::ChildFaulted,
-                Reason::ByClient => rpc::ChildState::ChildFaulted,
-                Reason::AdminCommandFailed => rpc::ChildState::ChildFaulted,
-            },
-        }
+fn map_fault_reason(r: FaultReason) -> ChildStateReason {
+    use ChildStateReason::*;
+
+    match r {
+        FaultReason::Unknown => None,
+        FaultReason::CantOpen => CannotOpen,
+        FaultReason::NoSpace => NoSpace,
+        FaultReason::TimedOut => TimedOut,
+        FaultReason::IoError => IoFailure,
+        FaultReason::Offline => ByClient,
+        FaultReason::RebuildFailed => RebuildFailed,
+        FaultReason::AdminCommandFailed => AdminFailed,
+        FaultReason::OfflinePermanent => ByClient,
     }
 }
-impl From<&NexusChild<'_>> for rpc::ChildStateReason {
-    fn from(child: &NexusChild) -> Self {
-        if child.is_opened_unsync() {
-            return Self::OutOfSync;
-        }
 
-        match child.state() {
-            ChildState::Init => Self::Init,
-            ChildState::ConfigInvalid => Self::ConfigInvalid,
-            ChildState::Open => Self::None,
-            ChildState::Destroying => Self::Closed,
-            ChildState::Closed => Self::Closed,
-            ChildState::Faulted(reason) => match reason {
-                Reason::NoSpace => Self::NoSpace,
-                Reason::TimedOut => Self::TimedOut,
-                Reason::Unknown => Self::None,
-                Reason::CantOpen => Self::CannotOpen,
-                Reason::RebuildFailed => Self::RebuildFailed,
-                Reason::IoError => Self::IoFailure,
-                Reason::ByClient => Self::ByClient,
-                Reason::AdminCommandFailed => Self::AdminFailed,
+fn map_child_state(child: &NexusChild) -> (ChildState, ChildStateReason) {
+    use ChildState::{
+        ChildDegraded as Degraded,
+        ChildFaulted as Faulted,
+        ChildOnline as Online,
+        ChildUnknown as Unknown,
+    };
+    use ChildStateReason::*;
+
+    match child.state_client() {
+        ChildStateClient::Init => (Degraded, Init),
+        ChildStateClient::ConfigInvalid => (Faulted, ConfigInvalid),
+        ChildStateClient::Open => (Online, None),
+        ChildStateClient::Closed => (Degraded, Closed),
+        ChildStateClient::Faulted(r) => (
+            match r {
+                FaultReason::IoError => Faulted,
+                s if s.is_recoverable() => Degraded,
```

Comment:
You've set IoError as is_recoverable, but here it's Faulted, is it to keep back-compat for now?

Reply:
Yes it is for back-compat, for now. We need to discuss this https://dcsw-bolt.atlassian.net/browse/GTM-469, and come up with better child states.